src/MagnumPlugins/BasisImageConverter/BasisImageConverter.cpp

```diff
@@ -0,0 +1,216 @@
+    basisu::job_pool jpool(threadCount);
+    params.m_pJob_pool = &jpool;
+
+#define PARAM_CONFIG(name, type) params.m_##name = configuration().value<type>(#name)
+    PARAM_CONFIG(compression_level, int);
+
+    PARAM_CONFIG(y_flip, bool);
+    PARAM_CONFIG(debug, bool);
+    PARAM_CONFIG(debug_images, bool);
+    PARAM_CONFIG(global_sel_pal, bool);
+    PARAM_CONFIG(no_auto_global_sel_pal, bool);
+    PARAM_CONFIG(perceptual, bool);
+    PARAM_CONFIG(no_endpoint_rdo, bool);
+    PARAM_CONFIG(no_selector_rdo, bool);
+    PARAM_CONFIG(read_source_images, bool);
+    PARAM_CONFIG(write_output_basis_files, bool);
+    PARAM_CONFIG(compute_stats, bool);
+    PARAM_CONFIG(check_for_alpha, bool);
+    PARAM_CONFIG(force_alpha, bool);
+    PARAM_CONFIG(multithreading, bool);
+    PARAM_CONFIG(seperate_rg_to_color_alpha, bool);
+    PARAM_CONFIG(disable_hierarchical_endpoint_codebooks, bool);
+
+    PARAM_CONFIG(hybrid_sel_cb_quality_thresh, float);
+
+    PARAM_CONFIG(global_pal_bits, int);
+    PARAM_CONFIG(global_mod_bits, int);
+
+    PARAM_CONFIG(endpoint_rdo_thresh, float);
+    PARAM_CONFIG(selector_rdo_thresh, float);
+
+    PARAM_CONFIG(mip_gen, bool);
+    PARAM_CONFIG(mip_renormalize, bool);
+    PARAM_CONFIG(mip_wrapping, bool);
+    PARAM_CONFIG(mip_srgb, bool);
+    PARAM_CONFIG(mip_scale, float);
+    PARAM_CONFIG(mip_smallest_dimension, int);
+    PARAM_CONFIG(mip_filter, std::string);
+
+    PARAM_CONFIG(max_endpoint_clusters, int);
+    PARAM_CONFIG(max_selector_clusters, int);
+    PARAM_CONFIG(quality_level, int);
+#undef PARAM_CONFIG
+
+    /* Without this, basis tries to find some global selector cookbook
+       and crashes on an assert that it does not exit */
+    params.m_no_hybrid_sel_cb = true;
+
+    /* Enable verbose debug output in basis to get reasons for failures */
+    basisu::enable_debug_printf(configuration().value<bool>("enable_debug_printf"));
+
+    if(image.size().x() <= 0 || image.size().y() <= 0) {
+        Error() << "Trade::BasisImageConverter::exportToData(): source image is empty";
+        return {};
+    }
+
+    /* Copy image data into the basis image. There is no way to construct a
+       basis image from existing data as it is based on a std::vector */
+    params.m_source_images.emplace_back(image.size().x(), image.size().y());
+
+    /* Get data properties and calculate the initial slice based on subimage
+       offset */
+    const std::pair<Math::Vector2<std::size_t>, Math::Vector2<std::size_t>> dataProperties =
+        image.dataProperties();
+    Containers::ArrayView<const char> inputData = image.data().suffix(
+        dataProperties.first.sum());
+
+    basisu::image& basisImage = params.m_source_images.back();
+
+    /* Do Y-flip and tight packing of image data */
+    const std::size_t rowSize = image.size().x()*image.pixelSize();
+    const std::size_t rowStride = dataProperties.second.x();
+    const std::size_t packedDataSize = rowSize*image.size().y();
```

Comment:
We discussed this on gitter quite a bit, but the additional flipping of the input data adds another nuance of course. I expect it to be like this:

 - If the Y flipped flag is set on the basis image, it should display correctly in OpenGL without need to change UVs.
 - If the Y flipped flag is not set, we need to flip.
 - Since basis expects the data to be "unflipped", but I assume that Magnum Images are flipped in general, because of OpenGL, we need to flip the magnum image data. Simply inverting the converter parameter is insufficient, because that would set the incorrect flag in the resulting basis image.

If my expectations are correct, the comment should stay and instead I should be flipping the image. @mosra What about point 2, is that correct? I'll check the others.

Reply:
Hah, you're right about the third point :see_no_evil: I assumed the Y flip was done in their frontend (`basisu_tool.cpp`) and the encoder just consumed the flag. That would be a way to do this properly, I think, since it would allow pre-flipped images to be put there. (Another PR flying their way, then.)

> instead I should be flipping the image

No, I think a better way would be to "post-patch" the file and add this flag to their header. Should be quite robust as long as we test properly.

For the second point -- for Vulkan, which has again Y at top AFAIK, I'll need to equip all importers with a configurable flag for Y flipping. So the final condition would be rather `if(fileYflipped != apiNeedsImageFlip) flip()`. But until then, yes, what you say is correct.

Reply:
> No, I think a better way would be to "post-patch" the file and add this flag to their header.

This will result in a bit of convoluted logic, because as the user I would try to produce the "y-flipped" flag by setting `y-flipped`.
Hence, if the user sets y-flipped, we need to pass the image to the basis converter as is and then set the y-flip flag. (Without actually passing `y-flipped` parameter to basis, to perserve the data as-is!)
But when the user requests non-y-flip flag, we need to enable y-flipped on the converter, so that it "unflips" the image, we would still be inverting the flag afterwards.

Reply:
Yeah, it would be basically passing the inverse of the Y-flip flag to the converter and then inverting the flag again on the resulting file.

.... until there's globally controllable Y-flip or the client-side flip I did for the transcoder it'll be kinda shitty like this, yeah.

Reply:
Actually, we have to copy every image row by row anyway, for RGB we need to re-pack to RGBA, RG => RRRG, R => RRR1, RGBA also needs to be copied because of skip and all.
Flipping of the input image is implemented at the current state of the PR.

Reply:
So what's the conclusion here? Still do the flag fixup rather than flipping the image on copy?

Reply:
Flag fixup, yeah. I'll submit a PR to Basis with this and it'll be easier to update that way.

Reply:
Because of the CRC/Checksum we decided to go with flipping on copy.